src/main/java/rx/observables/SyncOnSubscribe.java

```diff
@@ -0,0 +1,292 @@
+
+			}
+		};
+		return new LambdaOnSubscribe<S, T>(generator, nextFunc, onUnsubscribe);
+	}
+
+	/**
+	 * @param next
+	 * @param onUnsubscribe
+	 * @return
+	 */
+	@Experimental
+	public static <T> SyncOnSubscribe<Void, T> create(final Action1<Subscriber<? super T>> next,
+			final Action0 onUnsubscribe) {
+		Func0<Void> generator = new Func0<Void>() {
+			@Override
+			public Void call() {
+				return null;
+			}
+		};
+		Func2<Void, Subscriber<? super T>, Void> nextFunc = new Func2<Void, Subscriber<? super T>, Void>() {
+			@Override
+			public Void call(Void state, Subscriber<? super T> subscriber) {
+				next.call(subscriber);
+				return null;
+			}
+		};
+		Action1<Void> wrappedOnUnsubscribe = new Action1<Void>() {
+			@Override
+			public void call(Void t) {
+				onUnsubscribe.call();
+			}
+		};
+		return new LambdaOnSubscribe<Void, T>(generator, nextFunc, wrappedOnUnsubscribe);
+	}
+
+	@Override
+	public final void call(final Subscriber<? super T> subscriber) {
+		state = generateState();
+		final SyncOnSubscribe<S, T> self = this;
+		final Subscription s = Subscriptions.create(new Action0() {
+			@Override
+			public void call() {
+				self.onUnsubscribe(state);
+			}
+		});
+		subscriber.add(s);
+		Action1<Subscriber<? super T>> onNext = new Action1<Subscriber<? super T>>() {
+			@Override
+			public void call(Subscriber<? super T> subscriber) {
+				state = next(state, subscriber);
+			}
+		};
+		subscriber.setProducer(new SubscriptionProducer<T>(subscriber, onNext));
+	}
+
+	/**
+	 * An implementation that forwards the three main methods ({@code next},
+	 * {@code onSubscribe}, and {@code onUnsubscribe}) to functional callbacks.
+	 *
+	 * @param <T>
+	 *            the type of compatible Subscribers
+	 * @param <S>
+	 *            the type of the user-defined state
+	 */
+	private static final class LambdaOnSubscribe<S, T> extends SyncOnSubscribe<S, T> {
+		private final Func0<S> generator;
+		private final Func2<? super S, ? super Subscriber<? super T>, ? extends S> next;
+		private final Action1<? super S> onUnsubscribe;
+
+		private LambdaOnSubscribe(Func0<S> generator, Func2<? super S, ? super Subscriber<? super T>, ? extends S> next,
+				Action1<? super S> onUnsubscribe) {
+			this.generator = generator;
+			this.next = next;
+			this.onUnsubscribe = onUnsubscribe;
+		}
+
+		@Override
+		protected S generateState() {
+			return generator.call();
+		}
+
+		@Override
+		protected void onUnsubscribe(S state) {
+			onUnsubscribe.call(state);
+		}
+
+		@Override
+		protected S next(S state, Subscriber<? super T> subscriber) {
+			return next.call(state, subscriber);
+		}
+	}
+
+	/**
+	 * Contains the producer loop that reacts to downstream requests of work.
+	 *
+	 * @param <T>
+	 *            the type of compatible Subscribers
+	 */
+	private static final class SubscriptionProducer<T> implements Producer {
+		private final AtomicLong requestCount = new AtomicLong(0);
+		private final AtomicLong sentCount = new AtomicLong(0);
```

Comment:
sentCount appears to be accessed only in the emission loop so there is no need for an `AtomicLong` here or any other type besides a boolean since only 1 onNext is allowed ber callback.